### PR TITLE
[Clang][Sema]: Allow copy constructor side effects

### LINKED ISSUE #79518: `-Wunused` fails to realize copy initialization might have side effects
Compiling with `-Wunused-variable`, we expect no warning to be emitted (constructor calls might have side effects), yet a warning is emitted for `s2`

```cpp
struct S {
    S(int);
};

int main() {
    S s(0); // no warning
    S s2 = 0; // unused
    S s3{0}; // no warning
}
```

https://godbolt.org/z/6zzfE9c4a

### PR DESCRIPTION
**Pull request by vinayakdsci**
Fixes #79518

Copy constructors can have initialization with side effects, and thus clang should not emit a warning when -Wunused-variable is used in this context. Currently however, a warning is emitted.

Now, compilation happens without warnings.

### CODE REVIEW

**Comment by cor3ntin** [comment]
This change needs a release note.
Please add an entry to `clang/docs/ReleaseNotes.rst` in the section the most adapted to the change, and referencing any Github issue this change fixes. Thanks!

**Comment by vinayakdsci** [comment]
> This change needs a release note. Please add an entry to `clang/docs/ReleaseNotes.rst` in the section the most adapted to the change, and referencing any Github issue this change fixes. Thanks!

Done!

File: clang/test/SemaCXX/warn-unused-variables.cpp
Comment (vinayakdsci) [reply clang/test/SemaCXX/warn-unused-variables.cpp:2]: I think that would give an error when the gnu++11 check is run?

Comment (vinayakdsci) [reply clang/test/SemaCXX/warn-unused-variables.cpp:2]: > I think that would give an error when the gnu++11 check is run?

It seems to fail on all standards below C++14.

Comment (vinayakdsci) [reply clang/test/SemaCXX/warn-unused-variables.cpp:2]: That seems fair. How do you think I should proceed? Should I write a new test file?

Comment (cor3ntin) [reply clang/test/SemaCXX/warn-unused-variables.cpp:2]: I agree with @Endilll.
We should understand _why_ the behavior in older language modes is different.
I would recommend trying to use a debugger to try to see why we are not passing in that code path.
The difference in behavior across language modes looks like a bug that should be fixed, rather than a change in expected behavior

(in all language modes, if a constructor is called, that constructor might have side effects)

Comment (vinayakdsci) [reply clang/test/SemaCXX/warn-unused-variables.cpp:2]: Got it! I am on it, please give me some time and I'll report back.

Comment (vinayakdsci) [reply clang/test/SemaCXX/warn-unused-variables.cpp:2]: OK, the difference in behaviour seems to stem from the fact that in C++-11 mode and C++-14 mode, the constructors are marked as not elidable(clang/lib/Sema/SemaDecl.cpp, line 2049). This causes it to return false when these constructors which are created by copy initialization are encountered, specifically while passing in C++ standard modes lower than C++-17

**Comment by vinayakdsci** [comment]
@Endilll as you mention on the discourse, I have divided the tests as per the difference in the standards. The tests seem to be working fine. I hope you can review once you find time.

Thanks!

File: clang/test/SemaCXX/warn-unused-variables.cpp
Comment (Endilll) [inline clang/test/SemaCXX/warn-unused-variables.cpp:192]: Code around here can be replaced with a single line:
```
Foo fooScalar = 1; // cxx98-14-warning {{unused variable 'fooScalar'}}
```
Look at DR tests again how to make this happen (via `-verify=`):
https://github.com/llvm/llvm-project/blob/3902f9b6e2d925d50f9a4861d78e5aba07b6ef11/clang/test/CXX/drs/dr0xx.cpp#L1-L3

Comment (Endilll) [reply clang/test/SemaCXX/warn-unused-variables.cpp:192]: Here is documentation for `-verify` if you feel you need it: https://clang.llvm.org/docs/InternalsManual.html#verifying-diagnostics

**Comment by vinayakdsci** [comment]
I hope this does it!

**Review by Endilll — APPROVED** [review]
LGTM
Make sure @cor3ntin is happy, too.

**Comment by vinayakdsci** [comment]
> LGTM Make sure @cor3ntin is happy, too.

Thanks a lot! I have requested a review from @cor3ntin too. Thanks for your guidance!

**Comment by vinayakdsci** [comment]
@cor3ntin is this good to go?

Thanks!

**Comment by vinayakdsci** [comment]
@Endilll sorry for repeated pings, but I think @cor3ntin might be possibly busy.

If this looks good, can we go ahead with this PR?

Thanks!

**Comment by cor3ntin** [comment]
We have a policy against more than one ping per week, thanks.

FYI, if we are not going to fix the C++11 case, we should be very explicit about it. IE this PR is not a complete fix of https://github.com/llvm/llvm-project/issues/79518

**Comment by vinayakdsci** [comment]
Sorry about the pings, I think I got a little ahead of myself there.
But for C++11, we got to the conclusion that copy elision gives different behaviour for C++17 and later and thus different diagnostics?
In case I made a mistake, I'd be very happy to fix it.

Thanks
P.S. Sorry about the pings, please forgive me.

**Comment by vinayakdsci** [comment]
@cor3ntin gentle ping. Could you direct on what other changes are required?

Thanks!

**Comment by vinayakdsci** [comment]
OK, it looks like there should not be a warning in C++11 and 14 too, as copy initialization with side effects appears to be permitted sine C++11.
Could you direct me to the functions that Clang uses internally for implicit type conversion?
I am sure that using that I will be able to handle the other cases too.

Thanks!

**Comment by vinayakdsci** [comment]
@cor3ntin gentle ping. How exactly should I proceed? copy elision was made mandatory in C++17, so wouldn't a warning for C++ < 17 be allowable?

I am very happy to implement any suggestions you have for this PR, but I am unable to get the code to work for C++ < 17 because the constructors are not marked elidable in the code for these standards.

Specifically in the case where the class object is initialized by an '=' sign instead of an explicit argument being passed. As the RHS of the '=' is different from a `CXXRecordDecl`, `isTemporaryObject()` returns false and thus sets the constructor as non-elidable. This is where I am not sure on how to proceed, as changing this behavior breaks a number of assertions in the code path followed while compiling the source program. Either I could check if the RHS of the assignment is an rvalue, and then add an OR condition to set Elidable for the constructor to true. However, _whenever_ the constructor is elidable, the codegen emits _directly_, and expects an aggregate type to passed in, which is not true for a scalar such as an `int`.

Also, this would involve a conversion process in the code?

I would really appreciate pointers on how to solve this problem, and how to improve this PR too.

Thanks!

**Comment by vinayakdsci** [comment]
@AaronBallman does this PR look good to go, or should I make any changes and updates? I would really be grateful for any pointers on how this can be improved.

Thanks a lot!

**Review by AaronBallman — COMMENTED** [review]
I think we should handle pre-C++17 as well, because the constructors *are* called there: https://godbolt.org/z/YWjb8P67c (the ASTs are different but the runtime effects are not).

**Comment by AaronBallman** [comment]
> @AaronBallman C++17 and later mark all of these constructor calls as _non_-elidable after ignoring the implicit casts, while C++14 and before mark the constructor that is assigned by `=` as elidable, and hence the warning. Should the constructor also be marked as elidable in C++ 17 and later, or not marked as elidable in C++14 and earlier?

I just looked into the code and now I think it is reasonable to leave the C++14 and earlier behavior alone on the assumption that it's good enough for now and because the C++17 and up changes are an improvement. Whether a constructor is or isn't elidable is complex and there's a fair number of FIXME comments around it, so I don't think we should tie that to this PR.

**Review by AaronBallman — APPROVED** [review]
Aside from some grammatical changes to the release note, the changes LGTM! Do you need someone to land this on your behalf once those changes are made?

**Comment by vinayakdsci** [comment]
> Aside from some grammatical changes to the release note, the changes LGTM! 

Thank you so much! I have made changes to the Release Notes as you suggested. Thanks for your review!

> Do you need someone to land this on your behalf once those changes are made?

Yes, I don't have commit access at present, so I would really appreciate it if you or someone else could land it on my behalf.

Again, a big thanks!

**Pull request merged**
